cmd/virt-handler/virt-handler.go

```diff
@@ -155,6 +166,20 @@ func (app *virtHandlerApp) Run() {
 	if err != nil {
 		panic(err)
 	}
+
+	app.markNodeAsUnschedulable(logger)
+
+	go func() {
+		sigint := make(chan os.Signal, 1)
+
+		signal.Notify(sigint, syscall.SIGTERM)
```

Comment:
We will hit the bug again if virt-handler will be killed.

Though this appraoch minimizes the chances for that to happen but we still can hit the bug.

Maybe we need to move the logic the marks nodes as un-schedulable to another component (virt-controller? :thinking: )
and do something like:

1. If there's a healthy virt-handler on that node, do nothing.
2. Otherwise, mark it as un-schedulable.

virt-handler will then be responsible to mark the node as schedulable instead of unschedulable.

Reply:
> We will hit the bug again if virt-handler will be killed.

Not sure we have a bug here. We can thing about faster recovery. But we will recover in the case that a bug or another serious issues in the cluster will cause a forced shutdown. We have a heartbeat right now from virt-handler. Once we are sure that virt-handler is not in control anymore, our node-controller (in the virt-controller binary already) will mark it as unschedulable.

> Maybe we need to move the logic the marks nodes as un-schedulable to another component (virt-controller?  )
> and do something like:
> 
> 1. If there's a healthy virt-handler on that node, do nothing.
> 2. Otherwise, mark it as un-schedulable.

On the one hand I like the idea of immediately marking the node as unschedulable if we see that no virt-handler is running on the node. There is then really no need to wait for the heartbeat to expire. On the other hand I like about the heartbeat check that it is a clear interface and that virt-controller does not interpret virt-handler. That makes things more complicated in case we re-architect parts of virt-handler.

Reply:
Alright, as you suggested - let's discuss it as part of future optimizations and see if the suggested approach here worth the work that it requires. Since we have the heartbeat mechanism, I'm not 100% sure that it worth it.